### PR TITLE
Fixed registration channel filters

### PR DESCRIPTION
**Pull request by rbatistadev**
#### Type Of Change 
- [X] Small bug fix
- [ ] New feature or enhancement
- [ ] Breaking change

#### Checklist:
- [X] Commits are named and have tag following [commit rules](https://github.com/irontec/ivozprovider/blob/bleeding/doc/dev/en/commits.md)
- [X] Commits are split per component (schema, web/admin, kamusers, agis, ..)
- [X] Changes have been tested locally
- [ ] Fixes an existing issue (Fixes #XXXX) 
- [ ] Upport from existing Pull request #XXXX

#### Description
Fixed RegistrationChannelResolver when clients send, ambiguous filtering

### CODE REVIEW

File: web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature [hunk 1]
```diff
@@ -11,6 +11,20 @@ Feature: Retrieve active calls filters
...
+      And I send a "GET" request to "/my/active_calls/realtime_filter?b=1&c=2&dp=4"
+     Then the response status code should be 200
+      And the response should be in JSON
+      And the header "Content-Type" should be equal to "application/json"
+      And the JSON should be equal to:
+      """
+      {
+          "criteria": "trunks:b1:c2:dp4:*"
       }
       """
```
Comment (Kaian) [inline web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature:30]: Can you also add tests for filters for a given Carrier?
Can you also add tests for DdiProvider&Direction and Carrier&Direction ? P.e. DdiProvider and direction = inbound and DdiProvider and direction = outbound.

This tests will also be interesting for the other APIs

**Review by Kaian — CHANGES_REQUESTED** [review]

File: web/rest/brand/features/my/activeCalls/getActiveCallsFilters.feature [hunk 2]
```diff
@@ -11,6 +11,76 @@ Feature: Retrieve active calls filters
...
+    And I send a "GET" request to "/my/active_calls/realtime_filter?cr=2"
+    Then the response status code should be 200
+    And the response should be in JSON
+    And the header "Content-Type" should be equal to "application/json"
+    And the JSON should be equal to:
+      """
+      {
+          "criteria": "trunks:b1:*:cr2:*"
+      }
+      """
```
Comment (Kaian) [inline web/rest/brand/features/my/activeCalls/getActiveCallsFilters.feature:58]: Duplicate test?

File: web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature [hunk 3]
```diff
@@ -11,6 +11,76 @@ Feature: Retrieve active calls filters
...
+    And I send a "GET" request to "/my/active_calls/realtime_filter?cr=2"
+    Then the response status code should be 200
+    And the response should be in JSON
+    And the header "Content-Type" should be equal to "application/json"
+    And the JSON should be equal to:
+      """
+      {
+          "criteria": "trunks:*:*:cr2:*"
+      }
+      """
```
Comment (Kaian) [inline web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature:58]: Also duplicated

**Review by Kaian — CHANGES_REQUESTED** [review]

**Comment by Kaian** [comment]
We also need additional tests for direction only filtering.
There trunks realtime keys have the format:

```
trunks:<brand tag>:<company tag>:<carrier tag|ddi provider tag>:<call-id>
```

Keys with carrier are **outbound** calls and calls with ddi provider are **indbound** calls.

If you want to get only outbound calls, we need a filter that matches all carriers, pe:

```
trunks:*:*:cr*:*
```

If you want to get only inbound calls, we need a filter that matches all ddi providers, pe:
```
trunks:*:*:dp*:*
```

Take into account that carrier and ddi provider filters should have more priority that direction. If you want outbound calls for carrier #1, the key should be:

```
trunks:*:*:cr1:*
```

In that case, the direction does not provide any extra filtering, because all calls from Carrier #1 are outbound. 

If the user supplies Carrier and Direction inbound (which makes no sense), you can give an error, just like filtering by Carrirer and DDI provider at the same time.

File: library/Ivoz/Provider/Application/Service/ActiveCall/RegistrationChannelResolver.php [hunk 4]
```diff
@@ -39,12 +39,15 @@ public function execute(?array $filters, array $tokenPayload): string
...
-            if ($direction === OUTBOUND && is_null($cr)) {
-                $dp = '*';
+            $canFilterByDirection = !is_null($direction) && is_null($cr) && is_null($dp);
+            if ($canFilterByDirection) {
+                if ($direction === 'inbound') {
+                    $cr = 'cr*';
+                }
+
+                if ($direction === 'outbound') {
+                    $dp = 'dp*';
```
Comment (Kaian) [inline library/Ivoz/Provider/Application/Service/ActiveCall/RegistrationChannelResolver.php:49]: Actually the other way around. DDI Providers are used for inbound calls, Carrier for outbound calls. If you filter for outbound call you want to match all carriers (`cr*`);

**Review by Kaian — CHANGES_REQUESTED** [review]

File: web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature [hunk 5]
```diff
@@ -11,6 +11,90 @@ Feature: Retrieve active calls filters
...
+      And I send a "GET" request to "/my/active_calls/realtime_filter?dp=1&direction=outbound"
+     Then the response status code should be 200
+      And the response should be in JSON
+      And the header "Content-Type" should be equal to "application/json"
+      And the JSON should be equal to:
+      """
+      {
+          "criteria": "trunks:*:*:dp1:*"
+      }
+      """
```
Comment (Kaian) [inline web/rest/platform/features/my/activeCalls/getActiveCallsFilters.feature:72]: I would return error in these cases: 

```
"criteria": "trunks:-1"
```

While filtering for DDI Provider 1 all calls will be inbound. If you add a direction outbound to the filter, no call should be displayed, but with this response the direction filter is being ignored and Inbound DDI Provider calls will be displayed. Same applies to Carrier and outbound calls: if both filters are set, no call should be displayed.

There was a reference to this in my note:

> If the filter has Carrier and Direction inbound or DDI Provider and Direction outbound (which makes no sense), you can give an error, just like filtering by Carrirer and DDI provider at the same time.

**Review by Kaian — CHANGES_REQUESTED** [review]

**Review by mmadariaga — APPROVED** [review]

**Review by Kaian — CHANGES_REQUESTED** [review]
Filter when no parameters are passed is not correct . Add this test to **platform** to verify it:

```
  Scenario: Retrieve platform active calls filter with no parameters
    Given I add Authorization header
     When I add "Accept" header equal to "application/json"
      And I send a "GET" request to "/my/active_calls/realtime_filter"
     Then the response status code should be 200
      And the response should be in JSON
      And the header "Content-Type" should be equal to "application/json"
      And the JSON should be equal to:
      """
      {
          "criteria": "trunks:*:*:*:*"
      }
      """

This also happens in brand API. Add this test to **brand** to verify it:

```
  Scenario: Retrieve brand active calls filter json with no parameters
    Given I add Brand Authorization header
     When I add "Accept" header equal to "application/json"
      And I send a "GET" request to "/my/active_calls/realtime_filter"
     Then the response status code should be 200
      And the response should be in JSON
      And the header "Content-Type" should be equal to "application/json"
      And the JSON should be equal to:
      """
      {
          "criteria": "trunks:b1:*:*:*"
      }
      """
```

**Review by Kaian — APPROVED** [review]

**Pull request merged**